File: ValidPara.java
class Solution {
    private boolean iscorrect(char a1, char a2){
        if(a1 == '{' && a2 == '}')
            return true;
        
        if(a1 == '[' && a2 == ']')
            return true;
        
        if(a1 == '(' && a2 == ')'){
            return true;
        }
        return false;
    }
    
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for(int i = 0; i < s.length(); i++){
            char ch = s.charAt(i);
            if(ch == '[' || ch == '{' || ch == '('){
                stack.add(ch);
            }
            if(ch == ']' || ch == '}' || ch == ')'){
                if(stack.isEmpty()){
                    return false;
                }else if(!iscorrect(stack.pop(), ch)){
                    return false;
                }
            }
        }
        
        if(stack.isEmpty())
            return true;
        else
            return false;
       
    }
}

File: valid_para.py
'''
# Time complexity - O(N) : N is no.of char in str
# Space Complexity: O(N) 
# Did this code successfully run on LeetCode?: Yes
# Problems faced while coding this:None
# Approach : Push the bracket when open and pop it when closed.
'''
class Solution:
    def isValid(self, s: str) -> bool:
        if s is None or len(s) == 0 or len(s) % 2 != 0:
            return False
        
        stack = []
        for c in s:
            if c == "(":
                stack.append(")")
                
            elif c == "{":
                stack.append("}")
                
            elif c == "[":
                stack.append("]")
            
            elif len(stack) == 0 or stack.pop() != c:
                return False
            
        if len(stack) != 0:
            return False
        
        return True